### ISSUE TITLE
[Wave Collect][QBO] Handle the error

### ISSUE THREAD

**Issue by hayata-suenaga** [Expensify/App#41042]
Warn the user of the error that happens when Journal Entry is selected as the out-of-pocket expense export destination while tax is enabled.

[There is already the warning text added](https://github.com/Expensify/App/blob/6acc90392e927e905ee26282e8b2995287362a36/src/languages/en.ts#L1928) in the App repo. We can use this text for the error message.

<img width="400" alt="Screenshot 2024-04-25 at 6 46 50 PM" src="https://github.com/Expensify/App/assets/98560306/b46df1c9-3657-4c85-bac3-3225e8221899">

**Comment by trjExpensify** [comment]
@teneeto is going to take this as QBO clean-up. @tienifr you can progress #41045.

**Comment by teneeto** [comment]
Hi, I'm Eto from Callstack - expert contributor group - and I would like to take care of this issue.

**Comment by trjExpensify** [comment]
I don't think the OP as written is incredibly clear on what needs to be done for this issue. I think it would be good to outline what's expected, and then show a series of the mocks for each step to make that clear. There's an error message in-line when the JE/Tax misconfig happens in production already, what's missing? 

Is it that there should be an RBR from the bottom tab > Accounting LHN > Export > etc etc all the way to where the error message is in-line? More broadly for errors in the accounting settings, is that pattern applied or does it need to be added for any in-line errors in the QBO accounting configuration?

@zanyrenney I'm going to assign this one to you for the time being to help clarify and firm up on what's needed here, as I couldn't make that out completely from [this part of the doc](https://docs.google.com/document/d/1LLrOYRGEb_k_Z6dttcS99zlbCNNdpD1QlXAbFJYbR7g/edit#bookmark=id.8w72mwgy858c).

**Comment by zanyrenney** [comment]
👋🏼  hey there! Thanks for assigning me Tom. 

Breaking down the desired behaviour in steps.

1. User sets their export option as Journal Entries. They do not have Tax enabled so all export options are available.
2. User then enables Tax on the workspace
3. Then, we need to show the RBR for on the export option selected which is Journal Entries (as per step 1)
4. The RBR needs to show the error text “Journal entry is not available when taxes are enabled. Please select a different export option.”
5. On the `Export as` page, we should show Vendor bills and check as the only two options. NOT Journal entries
6. There is no "default" so both will appear unselected but in the selection list, available for either to be selected.
7. Once the user has selected either VBs or check, the RBR and error text should be removed.

**Comment by zanyrenney** [comment]
> Is it that there should be an RBR from the bottom tab > Accounting LHN > Export > etc etc all the way to where the error message is in-line? 
We should mirror the pattern from the error handling doc -  [here](https://docs.google.com/document/d/1aQl-AK_F6IqszWYG2aEllTsD5x3h4b8Zu0XEG0LuQ08/edit) so show the RBR on workspace > accounting  export

> More broadly for errors in the accounting settings, is that pattern applied or does it need to be added for any in-line errors in the QBO accounting configuration?
Not sure what you mean here, we're showing the RBR on the QBO line in the accounting page > connections card. LMK what you mean specifically though!

**Comment by trjExpensify** [comment]
>Breaking down the desired behaviour in steps.
>
>1. User sets their export option as Journal Entries. They do not have Tax enabled so all export options are available.
>2. User then enables Tax on the workspace
>3. Then, we need to show the RBR for on the export option selected which is Journal Entries (as per step 1)
>4. The RBR needs to show the error text “Journal entry is not available when taxes are enabled. Please select a different export option.”
>5. On the Export as page, we should show Vendor bills and check as the only two options. NOT Journal entries
>6. There is no "default" so both will appear unselected but in the selection list, available for either to be selected.
>7. Once the user has selected either VBs or check, the RBR and error text should be removed.

You might have to walk me through this live. These steps don't make sense to me, I think they're actually impossible? As soon as you connect to QBO, you can't "enable tax on the workspace", the only way you can turn that feature on thereafter is by importing `Taxes` via the QBO integration. So in this scenario you've outlined, wouldn't this be:

1. Export as: Journal Entry
2. Taxes: not imported
3. We don't let them toggle on the Taxes import in the QBO settings because the `Journal Entry` export method is selected
4. Show hint text beneath the Taxes import page that explains why and that they need to pick either VBs or Check as an export method to import taxes

OR

1. Export as: Vendor bill (which will be the default set on initial connection as well)
2. Taxes: Imported
3. We don't let them choose "Journal Entry" in the "Export as" select list because taxes are imported via the integration
4. We show hint text beneath the Export as push row explaining why JEs aren't available and that they need to stop importing taxes to export as JEs. 


>More broadly for errors in the accounting settings, is that pattern applied or does it need to be added for any in-line errors in the QBO accounting configuration?

On this one, what I mean by it is that right now there isn't a "brick road" that leads to the specific field in error within the QBO integration pages. You have to go there to find it by yourself, so we should make sure that a solution is implemented for that. I.e 

- RBR starts on the "Settings" bottom tab
- Red dot on the Workspaces menu item
- Red dot on the Workspace row that contains an error
- Red dot on the `Accounting` menu item within that Workspace
- Red dot on the applicable `Export || Import || Advanced` menu item where the error is
- Red dot and error message on the applicable field in error

**Comment by zanyrenney** [comment]
Happy to chat live if that helps Tom! This design was from this conversation: [here](https://expensify.slack.com/archives/C036QM0SLJK/p1706701037384029)

Jason raised it in the design doc and this was the [solution proposed ](https://expensify.slack.com/archives/C036QM0SLJK/p1706717531288809?thread_ts=1706701037.384029&cid=C036QM0SLJK)

But -- maybe this is an oversight and thinking through the bug found last week to do with Taxes (and the toggling!) it's this journey we're actually trying to solve for: 

- Export as: Journal Entry
- Taxes: not imported
- We don't let them toggle on the Taxes import in the QBO settings because the Journal Entry export method is selected
- Show hint text beneath the Taxes import page that explains why and that they need to pick either VBs or Check as an export method to import taxes

**Comment by trjExpensify** [comment]
Okay, cool.. so I think these are the two cases that don't allow for the misconfiguration and then there's no in-line error message or anything like that. 

**Vendor bill (or Check) + taxes imported, can't export as Journal Entry**
> 1. Export as: Vendor bill (which will be the default set on initial connection as well)
> 2. Taxes: Imported
> 3. We don't let them choose "Journal Entry" in the "Export as" select list because taxes are imported via the integration
> 4. We show hint text beneath the Export as push row explaining why JEs aren't available and that they need to stop importing taxes to export as JEs.

<img width="437" alt="image" src="https://github.com/Expensify/App/assets/16232057/d7b69d5d-9146-40a7-9cdf-00bab69130e0">

**Journal Entry +  Taxes not imported, can't import taxes**
>1. Export as: Journal Entry
>2. Taxes: not imported
>3. We don't let them toggle on the Taxes import in the QBO settings because the Journal Entry export method is selected
>4. Show hint text beneath the Taxes import page that explains why and that they need to pick either VBs or Check as an export method to import taxes

<img width="439" alt="image" src="https://github.com/Expensify/App/assets/16232057/ace7521e-c45e-4801-9433-9f7ea78494a7">

CC: @JmillsExpensify

**Comment by trjExpensify** [comment]
@zanyrenney @teneeto does the above make sense to you both on how we handle this?

**Comment by teneeto** [comment]
Hmmm, @trjExpensify I'll have to sink it in. but sounds good so far.

**Comment by trjExpensify** [comment]
Okay, let me know in the morning. It would be great to get this one over the line tomorrow!

**Comment by zanyrenney** [comment]
> @zanyrenney @teneeto does the above make sense to you both on how we handle this?
Yeah, this looks good. 

I think this is good because it shows the two pages (Taxes and Export As) as related and the settings within each as kinda "supporting" the other one. So, great thinking here!

I read back over the thread and feedback, and I think what we were trying to solve before was the user taking actions which led to misconfiguration and showing those with the RBR, but this feels like its more fool-proof in locking tax and also only showing those 2 export options. 

@trjExpensify @teneeto  let's do it! 🚀

**Comment by rojiphil** [comment]
> 4. We show hint text beneath the Export as push row explaining why JEs aren't available and that they need to stop importing taxes to export as JEs.
> 
> <img alt="image" width="437" src="https://private-user-images.githubusercontent.com/16232057/326577074-d7b69d5d-9146-40a7-9cdf-00bab69130e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NTc0MDcsIm5iZiI6MTcxNDU1NzEwNywicGF0aCI6Ii8xNjIzMjA1Ny8zMjY1NzcwNzQtZDdiNjlkNWQtOTE0Ni00MGE3LTljZGYtMDBiYWI2OTEzMGUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDA5NTE0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4MmFmMGU1NzM2MmViMDg0ODE4OTYxNDg2YmE0OTkyNGQ3ZDZlMzkwMWZiZGUzNjViMzU0ZjlhMDA4YTY2NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pbfTj_3E3iUFUZ8U38HqrBwPdgeFHES_9T-gn7L_pqA">

Just one small suggestion here. For the hint text here, we can reuse `outOfPocketTaxEnabledDescription` [here](https://github.com/Expensify/App/blob/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/languages/en.ts#L1972) which has a similar worded hint text i.e. 

`                outOfPocketTaxEnabledDescription: "Note: QuickBooks Online doesn't support a field for tax on Journal Entry exports. Because you have tax tracking enabled on your workspace, this export option is unavailable.",`

**Comment by rojiphil** [comment]
> **Journal Entry + Taxes not imported, can't import taxes**
> 
> 1. Export as: Journal Entry
> 2. Taxes: not imported
> 3. We don't let them toggle on the Taxes import in the QBO settings because the Journal Entry export method is selected
> 4. Show hint text beneath the Taxes import page that explains why and that they need to pick either VBs or Check as an export method to import taxes
> 
> <img alt="image" width="439" src="https://private-user-images.githubusercontent.com/16232057/326577294-ace7521e-c45e-4801-9433-9f7ea78494a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NTc0MDcsIm5iZiI6MTcxNDU1NzEwNywicGF0aCI6Ii8xNjIzMjA1Ny8zMjY1NzcyOTQtYWNlNzUyMWUtYzQ1ZS00ODAxLTk0MzMtOWY3ZWE3ODQ5NGE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDA5NTE0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxY2Q0ZGNjM2NmN2UxZmUxODBlNjE5NGQwNmJlYjVhYjBhZDE2ODE2YzYxMDUwNTRjYTkyZDRjMmMxYzI2MGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OSiDokPcpcCRmC0bc344T4tl2KccIzyFHdRGE7adbtY">

I am not able to see the QuickBooks taxes page for import. What am I missing here?

https://github.com/Expensify/App/assets/3004659/7dd5c379-946d-42d7-85e5-db39ebe5e801

**Comment by trjExpensify** [comment]
>Just one small suggestion here. For the hint text here, we can reuse outOfPocketTaxEnabledDescription [here](https://github.com/Expensify/App/blob/573d9893ae2a8e6e4ebe64088cb2fbf62ce81804/src/languages/en.ts#L1972) which has a similar worded hint text i.e.
>
>                outOfPocketTaxEnabledDescription: "Note: QuickBooks Online doesn't support a field for tax on Journal Entry exports. Because you have tax tracking enabled on your workspace, this export option is unavailable.",

That message doesn't make sense when you're looking at a toggle preventing you from turning on tax tracking...

**Comment by trjExpensify** [comment]
>I am not able to see the QuickBooks taxes page for import. What am I missing here?

You're likely connected to a instance of QuickBooks with a `US` country, whereby Tax isn't a feature and thus `Taxes` don't show as available for import. To test this, you'll need to connect to a `UK` entity.

**Comment by rojiphil** [comment]
> You're likely connected to a instance of QuickBooks with a `US` country, whereby Tax isn't a feature and thus `Taxes` don't show as available for import. To test this, you'll need to connect to a `UK` entity.

Ah! That makes sense. Let me try connecting to a `UK` entity.

**Comment by rojiphil** [comment]
> That message doesn't make sense when you're looking at a toggle preventing you from turning on tax tracking...

Oh! Did you mean the hint text on `Taxes` page? I was referring to the hint message on `Export as` page where the `Journal entry` option is hidden.  

`Note: QuickBooks Online does not support a field for tax on Journal Entry exports, as you have tax tracking enabled on your workspace, this export option is unavailable`
vs
`Note: QuickBooks Online doesn't support a field for tax on Journal Entry exports. Because you have tax tracking enabled on your workspace, this export option is unavailable.`

Both looked similar wording to me. no?

**Comment by zanyrenney** [comment]
Yeah I agree with @rojiphil @trjExpensify - maybe i am missing something in what you're saying but they seem like we could reuse that existing wording.

**Comment by zanyrenney** [comment]
Yeah I agree with @rojiphil we have that on Export as for JEs 

@trjExpensify - maybe i am missing something in what you're saying but they seem like we could reuse that existing wording.

**Comment by rojiphil** [comment]
Connecting to QuickBooks UK entity worked fine. I can now see the `Taxes` page. Thanks @trjExpensify

**Comment by trjExpensify** [comment]
>Oh! Did you mean the hint text on Taxes page? I was referring to the hint message on Export as page where the Journal entry option is hidden.

Ah right, yes, I thought you were suggesting to use that on the `Taxes` page as that's what is "new" here, and also where it doesn't make sense to re-use it. On the `Export as` page we should use what's in prod already -- looks like punctuation differences from Figma. 👍

**Comment by rojiphil** [comment]
@trjExpensify Since the solution approach used here for `Taxes` and `Journal Entry` option is better than the earlier one, we may want to create another issue to adopt the same approach for `Locations` and `Vendor bill/Check`. What do you think?

**Comment by hayata-suenaga** [comment]
> @trjExpensify Since the solution approach used here for Taxes and Journal Entry option is better than the earlier one, we may want to create another issue to adopt the same approach for Locations and Vendor bill/Check. What do you think?

I checked, and we're currently [not disabling Locations](https://github.com/Expensify/App/blob/cca1952e18ea67a587d25d8ff5064268b9425758/src/pages/workspace/accounting/qbo/import/QuickbooksLocationsPage.tsx#L47-L58) or hiding Vendor Bill or Check as reimbursable account options when Location import is enabled.

So we have to do this. I'll work on this.

**Comment by trjExpensify** [comment]
>@trjExpensify Since the solution approach used here for Taxes and Journal Entry option is better than the earlier one, we may want to create another issue to adopt the same approach for Locations and Vendor bill/Check. What do you think?

Yep, good idea! I've put a couple of suggestions in the slack thread [here](https://expensify.slack.com/archives/C036QM0SLJK/p1714757743749539?thread_ts=1714400674.229349&cid=C036QM0SLJK). @hayata-suenaga can move them into GH and an issue if he's happy!

**Comment by rojiphil** [comment]
@trjExpensify @hayata-suenaga That's awesome. And I would be happy to contribute as a reviewer there.

**Comment by hayata-suenaga** [comment]
@rojiphil this is [the issue](https://github.com/Expensify/App/issues/41621) and there is [a draft PR](https://github.com/Expensify/App/pull/41638) for it. There is a chance that [this other ongoing PR](https://github.com/Expensify/App/pull/41627) is enough to implement the prevention of invalid configuration options regarding locations and export account type selections.

**Comment by trjExpensify** [comment]
[PR](https://github.com/Expensify/App/pull/41638) is in review.

**Comment by trjExpensify** [comment]
@hayata-suenaga should we close this in favour of https://github.com/Expensify/App/issues/41621?

**Comment by hayata-suenaga** [comment]
Just as a clarification:

My PR is about location configuration

This issue is for the PR that handles tax configuration (which was done by Eto) 😄

**Comment by bfitzexpensify** [comment]
Adding another BZ team member to handle anything required on this while I'm OOO from today until June 11th:

@sonialiap - the PR - https://github.com/Expensify/App/pull/41638 - is in review, and we'll just need to follow the usual BZ process once it's deployed.

**Comment by hayata-suenaga** [comment]
yes we just need a payment for @s77rt as a C+

**Comment by rojiphil** [comment]
> yes we just need a payment for @s77rt as a C+ after the regression period is over

@hayata-suenaga Some confusion here. I was the C+ reviewer [here](https://github.com/Expensify/App/pull/41520#issuecomment-2112198759). Also reference [here](https://github.com/Expensify/App/pull/41520#issuecomment-2112198759). 
cc @trjExpensify @sonialiap

**Comment by hayata-suenaga** [comment]
ah sorry my bad I tagged a wrong C+ 🙇

**Comment by sonialiap** [comment]
[PR deployed to production](https://github.com/Expensify/App/pull/41520#issuecomment-2133552143) on May 27. +7 days puts payment on Jun 3

Payment summary:
- @rojiphil $250 - paid ✔️ 
- @s77rt $250 - offer sent

**Comment by s77rt** [comment]
@sonialiap I'm not due payment here. We can close the issue
